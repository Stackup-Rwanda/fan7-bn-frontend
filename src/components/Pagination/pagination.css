.body_display_pagination_items {
  height: auto;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
      -ms-flex-direction: row;
          flex-direction: row;
  width: 42%;
  margin: 0 auto;
  float: none;
}

.body_display_pagination_items_currentPage {
  margin-left: 25px;
  padding-top: 15px;
}

.body_display_pagination_items_rowsPerPage {
  margin-left: 25px;
  padding-top: 15px;
}

.body_display_pagination_items_allPages {
  margin-left: 25px;
  padding-top: 15px;
}

.body_display_pagination_items_nextPrev {
  padding-top: 15px;
  margin-left: 25px;
}
/*# sourceMappingURL=pagination.css.map */