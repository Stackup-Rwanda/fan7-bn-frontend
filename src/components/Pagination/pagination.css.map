{
    "version": 3,
    "mappings": "AAAA,AAAA,8BAA8B,CAAA;EAC1B,MAAM,EAAE,IAAI;EACZ,OAAO,EAAE,IAAI;EACb,cAAc,EAAE,GAAG;EACnB,KAAK,EAAE,GAAG;EACV,MAAM,EAAE,MAAM;EACd,KAAK,EAAE,IAAI;CAqBd;;AAnBI,AAAD,0CAAa,CAAA;EACT,WAAW,EAAE,IAAI;EACjB,WAAW,EAAE,IAAI;CACpB;;AAEA,AAAD,0CAAa,CAAA;EACT,WAAW,EAAE,IAAI;EACjB,WAAW,EAAE,IAAI;CACpB;;AAEA,AAAD,uCAAU,CAAA;EACN,WAAW,EAAE,IAAI;EACjB,WAAW,EAAE,IAAI;CACpB;;AAEA,AAAD,uCAAU,CAAA;EACN,WAAW,EAAE,IAAI;EACjB,WAAW,EAAE,IAAI;CACpB",
    "sources": [
        "pagination.scss"
    ],
    "names": [],
    "file": "pagination.css"
}