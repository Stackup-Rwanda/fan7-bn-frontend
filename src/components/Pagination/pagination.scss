.body_display_pagination_items{
    height: auto;
    display: flex;
    flex-direction: row;
    width: 42%;
    margin: 0 auto;
    float: none;

    &_currentPage{
        margin-left: 25px;
        padding-top: 15px;
    }
    
    &_rowsPerPage{
        margin-left: 25px;
        padding-top: 15px;
    }
    
    &_allPages{
        margin-left: 25px;
        padding-top: 15px;
    }
    
    &_nextPrev{
        padding-top: 15px;
        margin-left: 25px;
    }
}