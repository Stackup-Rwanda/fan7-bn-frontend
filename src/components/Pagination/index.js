import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './pagination.scss';

class Pagination extends Component {
    constructor(props) {
        super(props);
        this.state = {
            totalRecords: "",
            pageLimit: "",
            totalPages: "",
            currentPage: "",
            initialPage: "",
            pagesToShow: ""
        };
    }

    componentDidMount() {
        this.setState({
            totalRecords: this.props.totalRecords,
            pageLimit: this.props.pageLimit || 10,
            totalPages: Math.ceil(this.props.totalRecords / this.props.pageLimit),
            pagesToShow: this.props.pagesToShow || 5,
            currentPage: this.props.initialPage || 1
        });
    }

    componentWillReceiveProps(nextProps) {
        this.setState({
            totalRecords: nextProps.totalRecords,
            pageLimit: nextProps.pageLimit || 10,
            totalPages: Math.ceil(nextProps.totalRecords / nextProps.pageLimit),
            pagesToShow: nextProps.pagesToShow || 5
        });
    }

    componentDidUpdate(prevProps, prevState) {
        if (
            this.state.totalRecords !== prevState.totalRecords ||
            this.state.pageLimit !== prevState.pageLimit
        ) {
            this.setPage(this.state.currentPage);
        }
    }

    setPage(page) {
        const { totalRecords, pageLimit, totalPages } = this.state;

        if (page < 1) {
            page = 1;
        } else if (page > totalPages) {
            page = totalPages;
        }

        this.setState({
            currentPage: page
        });

        const startIndex = (page - 1) * pageLimit;
        const endIndex = Math.min(startIndex + pageLimit - 1, totalRecords - 1);

        this.props.onChangePage({
            pageLimit,
            totalPages,
            page,
            startIndex,
            endIndex
        });
    }

    getPager() {
        const {  currentPage, totalPages } = this.state;
        const pages = [];
           let startFromNumber;
           let pagesToShow = this.state.pagesToShow;
        if (totalPages <= pagesToShow) {
            startFromNumber = 1;
            pagesToShow = totalPages;
        } else {
            if (currentPage <= Math.ceil(pagesToShow / 2)) {
                startFromNumber = 1;
            } else if (
                currentPage + Math.floor((pagesToShow - 1) / 2) >=
                totalPages
            ) {
                startFromNumber = totalPages - (pagesToShow - 1);
            } else {
                startFromNumber = currentPage - Math.floor(pagesToShow / 2);
            }
        }

        for (let i = 1; i <= pagesToShow; i++) {
            pages.push(startFromNumber++);
        }

        return {
            currentPage,
            totalPages,
            pages
        };
    }

    render() {
        if (!this.state.totalRecords || this.state.totalPages === 1) return null;
        const {
            totalPages,
            currentPage,
            pageLimit,
          } = this.state;
        const pager = this.getPager();
        
        return (
            <div className="body_display_pagination_items">
                <div className="body_display_pagination_items_currentPage">
                    <p>
                        Page {currentPage}
                    </p>
                </div>

                <div className="body_display_pagination_items_rowsPerPage">
                    Rows per page
                        <select
                        className="form-control"
                        value={pageLimit}                        
                        onChange={e =>
                            this.setState({ pageLimit: parseInt(e.target.value) })
                        }>                            
                        <option value={5}>5</option>
                        <option value={10}>10</option>
                        <option value={25}>25</option>
                        <option value={50}>50</option>
                    </select>
                </div>
                <div className="body_display_pagination_items_allPages">
                    <p>
                        {currentPage} - {pageLimit} of {totalPages}
                    </p>
                </div>
                <div className="body_display_pagination_items_nextPrev">
                    <button
                        className="body_display_pagination_items_nextPrev_prev"
                        disabled={pager.currentPage === 1 ? true : false}
                        onClick={() => this.setPage(pager.currentPage - 1)}>
                        Prev
                    </button>
                    <button
                        className="body_display_pagination_items_nextPrev_next"
                        disabled={pager.currentPage === pager.totalPages ? true : false}
                        onClick={() => this.setPage(pager.currentPage + 1)}>
                        Next
                    </button>
                </div>

            </div>
        );
    }
}

Pagination.propTypes = {
    totalRecords: PropTypes.number.isRequired,
    pageLimit: PropTypes.number,
    initialPage: PropTypes.number,
    pagesToShow: PropTypes.number,
    onChangePage: PropTypes.func
};

export default Pagination;

