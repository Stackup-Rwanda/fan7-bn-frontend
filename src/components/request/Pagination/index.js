import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './Pagination.scss';
import caret from '../../../assets/icons/icons8-chevron-left-30.png';

class Pagination extends Component {
  constructor(props) {
    super(props);

    this.state = {
      total: undefined,
      page: 1,
      rowsPerPage: [],
      numberOfRows: 5,
    };

    this.selectRowsPerPage = this.selectRowsPerPage.bind(this);
    this.selectPageNumber = this.selectPageNumber.bind(this);
    this.renderRowsPerPage = this.renderRowsPerPage.bind(this);
    this.renderRowRange = this.renderRowRange.bind(this);
    this.numberOfPages = this.numberOfPages.bind(this);
    this.incrementPage = this.incrementPage.bind(this);
    this.decrementPage = this.decrementPage.bind(this);
  }

  // componentDidMount() {
  //   const { total, page, rowsPerPage, numberOfRows } = this.props;
  //   this.setState({
  //     total,
  //     page,
  //     rowsPerPage,
  //     numberOfRows,
  //   });
  // }

  selectRowsPerPage(event) {
    const { value } = event.target;
    const { total, page, onChangeRowsPerPage } = this.props;
    if (parseInt(value, 10) * page > total) {
      const updatedPage = Math.ceil(total / parseInt(value, 10), 10);
      onChangeRowsPerPage(parseInt(value, 10), updatedPage)
    } else {
      console.log(value, page);
      onChangeRowsPerPage(parseInt(value, 10), page)
    }
  }

  selectPageNumber(event) {
    const { value } = event.target;
    const { onChangePage, page,  ...rest } = this.props;
    onChangePage(event, value);
  }

  numberOfPages() {
    const { total, numberOfRows } = this.props;
    let numPages = [];
    for (let i = 0; i < Math.ceil(total / numberOfRows); i++) {
      numPages.push(i + 1);
    }

    return numPages.map((pageValue, index) => (
      <option key={index} value={pageValue}>
        {pageValue}
      </option>
    ));
  }

  incrementPage(event) {
    const { page, onChangePage, ...rest } = this.props;
    onChangePage(event, page + 1);
  }

  decrementPage(event) {
    const { page, onChangePage, ...rest } = this.props;
    onChangePage(event, page - 1);
  }

  renderRowsPerPage() {
    const { rowsPerPage } = this.props;
    return rowsPerPage.map((rowValue, index) => (
      <option key={index} value={rowValue}>
        {rowValue}
      </option>
    ));
  }

  renderRowRange() {
    const { total, page, numberOfRows } = this.props;
    return (
      <span>
        {total === 0 ? 0 : numberOfRows * page - numberOfRows + 1} -{' '}
        {numberOfRows * page < total ? numberOfRows * page : total}
      </span>
    );
  }

  render() {
    const { total, page, numberOfRows } = this.props;
    return (
      <div className="pagination-container">
        <div className="pagination-section">
          <div className="pagination-section__text">Page:</div>
          <select
            className="pagination-section__select"
            name="page"
            defaultValue={page}
            onChange={this.selectPageNumber}
          >
            {total === 1 ? null : this.numberOfPages()}
          </select>
        </div>
        <div className="pagination-section">
          <div className="pagination-section__text">Rows Per Page:</div>
          <select
            className="pagination-section__select"
            name="rowsPerPage"
            defaultValue={numberOfRows}
            onChange={this.selectRowsPerPage}
          >
            {this.renderRowsPerPage()}
          </select>
        </div>
        <div className="pagination-section">
          <div className="pagination-section__text">
            {this.renderRowRange()} of {total}
          </div>
        </div>
        <div className="pagination-section">
          <button
            className="pagination-section__navigation"
            type="button"
            name="decrement"
            disabled={page <= 1}
            onClick={this.decrementPage}
          >
            <img className="pagination-section__navigation___left" src={caret} />
          </button>
          <button
            className="pagination-section__navigation"
            type="button"
            name="increment"
            disabled={page >= total / numberOfRows} //rowsPerPage
            onClick={this.incrementPage}
          >
            <img className="pagination-section__navigation___right" src={caret} />
          </button>
        </div>
      </div>
    );
  }
}

Pagination.propTypes = {
  total: PropTypes.number,
  page: PropTypes.number,
  numberOfRows: PropTypes.number,
  rowsPerPage: PropTypes.array,
  updateRows: PropTypes.func,
};

Pagination.defaultProps = {
  total: 20,
  page: 1,
  rowsPerPage: [5, 10, 15, 20],
  numberOfRows: 5,
};

export default Pagination;
