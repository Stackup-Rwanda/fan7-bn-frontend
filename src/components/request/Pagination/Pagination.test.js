/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount, shallow } from 'enzyme';
import Pagination from './index';

describe('Pagination Component Test', () => {
  it('should return true if component exists', () => {
    const wrapper = shallow(<Pagination />);

    expect(wrapper.find('.pagination-container').exists()).toBe(true);
  });

  it('should handle selectPageNumber', () => {
    const onChangeRowsPerPage = jest.fn();
    const onChangePage = jest.fn();
    const props = {
      total: 20,
      rowsPerPage: [5, 10, 15, 20],
      page: 1,
      numberOfRows: 5,
      onChangeRowsPerPage,
      onChangePage,
    };

    const event = {
      target: { value: '2' },
    };

    const wrapper = shallow(<Pagination {...props} />);

    wrapper.find('select[name="page"]').simulate('change', event);

    expect(onChangePage).toHaveBeenCalled();
  });

  it('should handle selectRowsPerPage and not update page', () => {
    const onChangeRowsPerPage = jest.fn();
    const onChangePage = jest.fn();
    const props = {
      total: 20,
      rowsPerPage: [5, 10, 15, 20],
      page: 1,
      numberOfRows: 5,
      onChangeRowsPerPage,
      onChangePage,
    };
    const event = {
      target: { value: '10' },
    };

    const wrapper = shallow(<Pagination {...props} />);

    wrapper.find('select[name="rowsPerPage"]').simulate('change', event);

    expect(onChangeRowsPerPage).toHaveBeenCalled();
  });

  it('should handle selectRowsPerPage and update page', () => {
    const onChangeRowsPerPage = jest.fn();
    const onChangePage = jest.fn();
    const props = {
      total: 20,
      rowsPerPage: [5, 10, 15, 20],
      page: 2,
      numberOfRows: 5,
      onChangeRowsPerPage,
      onChangePage,
    };
    const event = {
      target: { value: '15' },
    };

    const wrapper = shallow(<Pagination {...props} />);

    wrapper.find('select[name="rowsPerPage"]').simulate('change', event);

    expect(onChangeRowsPerPage).toHaveBeenCalled();
  });

  it('should handle incrementPage', () => {
    const onChangeRowsPerPage = jest.fn();
    const onChangePage = jest.fn();
    const props = {
      total: 20,
      rowsPerPage: [5, 10, 15, 20],
      page: 1,
      numberOfRows: 5,
      onChangeRowsPerPage,
      onChangePage,
    };
    const event = {
      target: { value: '10' },
    };

    const wrapper = shallow(<Pagination {...props} />);

    wrapper.find('button[name="increment"]').simulate('click', event);

    expect(onChangePage).toHaveBeenCalled();
  });

  it('should handle decrementPage', () => {
    const onChangeRowsPerPage = jest.fn();
    const onChangePage = jest.fn();
    const props = {
      total: 20,
      rowsPerPage: [5, 10, 15, 20],
      page: 1,
      numberOfRows: 5,
      onChangeRowsPerPage,
      onChangePage,
    };
    const event = {
      target: { value: '10' },
    };

    const wrapper = shallow(<Pagination {...props} />);

    wrapper.find('button[name="decrement"]').simulate('click', event);

    expect(onChangePage).toHaveBeenCalled();
  });

  it('should render null if total equals to one', () => {
    const onChangeRowsPerPage = jest.fn();
    const onChangePage = jest.fn();
    const props = {
      total: 1,
      rowsPerPage: [5, 10, 15, 20],
      page: 1,
      numberOfRows: 5,
      onChangeRowsPerPage,
      onChangePage,
    };

    const wrapper = shallow(<Pagination {...props} />);

    expect(wrapper.find('select[name="page"]').text()).toEqual('');
  });
});

describe('<Pagination /> shallow rendering tests', () => {
  it('matches the snapshot', () => {
    const onChangeRowsPerPage = jest.fn();
    const onChangePage = jest.fn();
    const props = {
      total: 20,
      rowsPerPage: [5, 10, 15, 20],
      page: 1,
      numberOfRows: 5,
      onChangeRowsPerPage,
      onChangePage,
    };
    const tree = shallow(<Pagination {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
