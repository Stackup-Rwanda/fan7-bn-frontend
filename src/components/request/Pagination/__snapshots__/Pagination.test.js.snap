// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> shallow rendering tests matches the snapshot 1`] = `
<div
  className="pagination-container"
>
  <div
    className="pagination-section"
  >
    <div
      className="pagination-section__text"
    >
      Page:
    </div>
    <select
      className="pagination-section__select"
      defaultValue={1}
      name="page"
      onChange={[Function]}
    >
      <option
        key="0"
        value={1}
      >
        1
      </option>
      <option
        key="1"
        value={2}
      >
        2
      </option>
      <option
        key="2"
        value={3}
      >
        3
      </option>
      <option
        key="3"
        value={4}
      >
        4
      </option>
    </select>
  </div>
  <div
    className="pagination-section"
  >
    <div
      className="pagination-section__text"
    >
      Rows Per Page:
    </div>
    <select
      className="pagination-section__select"
      defaultValue={5}
      name="rowsPerPage"
      onChange={[Function]}
    >
      <option
        key="0"
        value={5}
      >
        5
      </option>
      <option
        key="1"
        value={10}
      >
        10
      </option>
      <option
        key="2"
        value={15}
      >
        15
      </option>
      <option
        key="3"
        value={20}
      >
        20
      </option>
    </select>
  </div>
  <div
    className="pagination-section"
  >
    <div
      className="pagination-section__text"
    >
      <span>
        1
         -
         
        5
      </span>
       of 
      20
    </div>
  </div>
  <div
    className="pagination-section"
  >
    <button
      className="pagination-section__navigation"
      disabled={true}
      name="decrement"
      onClick={[Function]}
      type="button"
    >
      <img
        className="pagination-section__navigation___left"
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </button>
    <button
      className="pagination-section__navigation"
      disabled={false}
      name="increment"
      onClick={[Function]}
      type="button"
    >
      <img
        className="pagination-section__navigation___right"
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </button>
  </div>
</div>
`;
